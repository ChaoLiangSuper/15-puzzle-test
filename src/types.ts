export interface IAppState {
  tiles: Number[];
  isSolved: boolean;
}

export interface ICellProp {
  displayValue: number;
  handleCellclick: () => void;
}

export interface IGridProp {
  tiles: number[];
  edgeLength: number;
  handleCellclick: (p: number) => void;
  isSolved: boolean;
}

export interface IOverlayProps {
  displayText: string;
}
