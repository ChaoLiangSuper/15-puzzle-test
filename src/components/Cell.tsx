import React from 'react';
import { ICellProp } from '../types';

const Cell: React.FC<ICellProp> = ({ displayValue, handleCellclick }) => {
  const className = displayValue === 0 ? 'cell cell-empty' : 'cell';
  return (
    <div className={className} onClick={handleCellclick}>
      {displayValue === 0 ? null : displayValue}
    </div>
  );
};

export default Cell;
