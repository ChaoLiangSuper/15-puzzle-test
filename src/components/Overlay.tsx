import React from 'react';
import { IOverlayProps } from '../types';

const Overlay: React.FC<IOverlayProps> = ({ displayText }) => (
  <div className='overlay'>{displayText}</div>
);

export default Overlay;
