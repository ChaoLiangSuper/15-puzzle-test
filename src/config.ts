export const edgeLength: number = 4;
