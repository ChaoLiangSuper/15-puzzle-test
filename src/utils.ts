import _ from 'lodash';

export const getNewTiles = (edgeLength: number) =>
  _(Array(edgeLength * edgeLength - 1))
    .map((value, index) => index + 1)
    .shuffle()
    .concat(0)
    .value();

export const getPosition = (position: number, edgeLength: number) => ({
  x: Math.floor(position / edgeLength),
  y: position % edgeLength,
});

export const isMoveable = (
  current: number,
  target: number,
  edgeLength: number
) => {
  const currentPosition = getPosition(current, edgeLength);
  const targetPosition = getPosition(target, edgeLength);

  if (
    (currentPosition.x === targetPosition.x &&
      Math.abs(currentPosition.y - targetPosition.y) === 1) ||
    (currentPosition.y === targetPosition.y &&
      Math.abs(currentPosition.x - targetPosition.x) === 1)
  ) {
    return true;
  }

  return false;
};

export const checkSolveState = (tiles: number[]) => {
  if (tiles[tiles.length - 1] === 0) {
    console.log(tiles.slice(0, tiles.length - 2));
    return tiles
      .slice(0, tiles.length - 2)
      .every((value, index) => value === index + 1);
  }
  return false;
};
