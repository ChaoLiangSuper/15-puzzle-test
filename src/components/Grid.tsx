import _ from 'lodash';
import React from 'react';
import Cell from './Cell';
import Overlay from './Overlay';
import { IGridProp } from '../types';

const Grid: React.FC<IGridProp> = ({
  tiles,
  edgeLength,
  handleCellclick,
  isSolved,
}) => {
  const cells = _.map(tiles, (value, index) => (
    <React.Fragment key={index}>
      <Cell
        displayValue={value}
        handleCellclick={() => handleCellclick(index)}
      />
      {(index + 1) % edgeLength === 0 ? <div /> : null}
    </React.Fragment>
  ));

  return (
    <div className='grid'>
      {cells}
      {isSolved ? <Overlay displayText='WIN' /> : null}
    </div>
  );
};

export default Grid;
