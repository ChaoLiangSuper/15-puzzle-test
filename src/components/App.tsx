import React from 'react';
import Grid from './Grid';
import { getNewTiles, isMoveable, checkSolveState } from '../utils';
import { edgeLength } from '../config';
import { IAppState } from '../types';

class App extends React.Component<{}, IAppState> {
  state = {
    tiles: getNewTiles(edgeLength),
    isSolved: false,
  };

  handleCellclick = (currentCell: number) => {
    const { tiles } = this.state;
    const targetCell = tiles.indexOf(0);
    if (isMoveable(currentCell, targetCell, edgeLength)) {
      const newTiles = [...tiles];
      const temp = newTiles[currentCell];
      newTiles[currentCell] = newTiles[targetCell];
      newTiles[targetCell] = temp;
      this.setState({
        tiles: newTiles,
        isSolved: checkSolveState(newTiles),
      });
    }
  };

  resetTiles = () => {
    this.setState({
      tiles: getNewTiles(edgeLength),
      isSolved: false,
    });
  };

  render() {
    const { tiles, isSolved } = this.state;
    return (
      <main className='background'>
        <Grid
          tiles={tiles}
          edgeLength={edgeLength}
          handleCellclick={this.handleCellclick}
          isSolved={isSolved}
        />
        <button className='button-reset' onClick={this.resetTiles}>
          RESET
        </button>
      </main>
    );
  }
}

export default App;
