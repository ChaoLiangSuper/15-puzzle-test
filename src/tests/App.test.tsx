import React from 'react';
import { shallow, mount, ReactWrapper, ShallowWrapper } from 'enzyme';
import App from '../components/App';
import Grid from '../components/Grid';
import Cell from '../components/Cell';
import Overlay from '../components/Overlay';
import * as utils from '../utils';
import { mockTiles, resetTiles } from './mock';
import { edgeLength } from '../config';
import { IAppState } from '../types';

describe('<App />', () => {
  let wrapper: ShallowWrapper<{}, IAppState>;
  let getMockTiles: jest.SpyInstance;

  beforeAll(() => {
    getMockTiles = jest
      .spyOn(utils, 'getNewTiles')
      .mockImplementation(() => mockTiles);
  });

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  afterAll(() => {
    getMockTiles.mockRestore();
  });

  it('should initialize without crash and match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should contains <Grid />', () => {
    const grid = wrapper.find(Grid);
    const gridProp = grid.props();
    expect(grid.length).toBe(1);
    expect(gridProp.edgeLength).toBe(edgeLength);
    expect(gridProp.isSolved).toBe(false);
    expect(gridProp.tiles).toBe(mockTiles);
  });

  it('should contains reset button', () => {
    const button = wrapper.find('button');
    getMockTiles.mockImplementation(() => resetTiles);
    expect(button.text()).toBe('RESET');
    button.simulate('click');
    expect(wrapper.state().tiles).toEqual(resetTiles);
  });
});

describe('Game Workflow', () => {
  let wrapper: ReactWrapper<{}, IAppState>;
  let getMockTiles: jest.SpyInstance;

  beforeAll(() => {
    getMockTiles = jest
      .spyOn(utils, 'getNewTiles')
      .mockImplementation(() => mockTiles);
  });

  beforeEach(() => {
    wrapper = mount(<App />);
  });

  afterAll(() => {
    getMockTiles.mockRestore();
  });

  it('should initialize without crash and match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should initialize with 15 cells and 1 empty cell', () => {
    const cells = wrapper.find(Cell);
    expect(cells.length).toBe(16);
    expect(cells.find('.cell').length).toBe(16);
    expect(cells.find('.cell.cell-empty').length).toBe(1);
  });

  it('should show win when the game is solved', () => {
    const cells = wrapper.find(Cell);
    cells.last().simulate('click');
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(Overlay).text()).toBe('WIN');
  });

  it('should be able to move tile vertically', () => {
    wrapper
      .find(Cell)
      .at(10)
      .simulate('click');
    expect(wrapper.find(Cell).map(cell => cell.props().displayValue)).toEqual([
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      0,
      12,
      13,
      14,
      11,
      15,
    ]);
  });

  it('should be able to move tile horizontally', () => {
    wrapper
      .find(Cell)
      .at(13)
      .simulate('click');
    expect(wrapper.find(Cell).map(cell => cell.props().displayValue)).toEqual([
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      0,
      14,
      15,
    ]);
  });

  it('should not be able to move non-moveable tile', () => {
    wrapper
      .find(Cell)
      .at(0)
      .simulate('click');
    expect(wrapper.find(Cell).map(cell => cell.props().displayValue)).toEqual([
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13,
      14,
      0,
      15,
    ]);
  });
});
