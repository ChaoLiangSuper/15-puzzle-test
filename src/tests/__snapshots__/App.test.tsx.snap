// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should initialize without crash and match snapshot 1`] = `
<main
  className="background"
>
  <Grid
    edgeLength={4}
    handleCellclick={[Function]}
    isSolved={false}
    tiles={
      Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
        9,
        10,
        11,
        12,
        13,
        14,
        0,
        15,
      ]
    }
  />
  <button
    className="button-reset"
    onClick={[Function]}
  >
    RESET
  </button>
</main>
`;

exports[`Game Workflow should initialize without crash and match snapshot 1`] = `
<App>
  <main
    className="background"
  >
    <Grid
      edgeLength={4}
      handleCellclick={[Function]}
      isSolved={false}
      tiles={
        Array [
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          0,
          15,
        ]
      }
    >
      <div
        className="grid"
      >
        <Cell
          displayValue={1}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            1
          </div>
        </Cell>
        <Cell
          displayValue={2}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            2
          </div>
        </Cell>
        <Cell
          displayValue={3}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            3
          </div>
        </Cell>
        <Cell
          displayValue={4}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            4
          </div>
        </Cell>
        <div />
        <Cell
          displayValue={5}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            5
          </div>
        </Cell>
        <Cell
          displayValue={6}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            6
          </div>
        </Cell>
        <Cell
          displayValue={7}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            7
          </div>
        </Cell>
        <Cell
          displayValue={8}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            8
          </div>
        </Cell>
        <div />
        <Cell
          displayValue={9}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            9
          </div>
        </Cell>
        <Cell
          displayValue={10}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            10
          </div>
        </Cell>
        <Cell
          displayValue={11}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            11
          </div>
        </Cell>
        <Cell
          displayValue={12}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            12
          </div>
        </Cell>
        <div />
        <Cell
          displayValue={13}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            13
          </div>
        </Cell>
        <Cell
          displayValue={14}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            14
          </div>
        </Cell>
        <Cell
          displayValue={0}
          handleCellclick={[Function]}
        >
          <div
            className="cell cell-empty"
            onClick={[Function]}
          />
        </Cell>
        <Cell
          displayValue={15}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            15
          </div>
        </Cell>
        <div />
      </div>
    </Grid>
    <button
      className="button-reset"
      onClick={[Function]}
    >
      RESET
    </button>
  </main>
</App>
`;

exports[`Game Workflow should show win when the game is solved 1`] = `
<App>
  <main
    className="background"
  >
    <Grid
      edgeLength={4}
      handleCellclick={[Function]}
      isSolved={true}
      tiles={
        Array [
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          10,
          11,
          12,
          13,
          14,
          15,
          0,
        ]
      }
    >
      <div
        className="grid"
      >
        <Cell
          displayValue={1}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            1
          </div>
        </Cell>
        <Cell
          displayValue={2}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            2
          </div>
        </Cell>
        <Cell
          displayValue={3}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            3
          </div>
        </Cell>
        <Cell
          displayValue={4}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            4
          </div>
        </Cell>
        <div />
        <Cell
          displayValue={5}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            5
          </div>
        </Cell>
        <Cell
          displayValue={6}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            6
          </div>
        </Cell>
        <Cell
          displayValue={7}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            7
          </div>
        </Cell>
        <Cell
          displayValue={8}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            8
          </div>
        </Cell>
        <div />
        <Cell
          displayValue={9}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            9
          </div>
        </Cell>
        <Cell
          displayValue={10}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            10
          </div>
        </Cell>
        <Cell
          displayValue={11}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            11
          </div>
        </Cell>
        <Cell
          displayValue={12}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            12
          </div>
        </Cell>
        <div />
        <Cell
          displayValue={13}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            13
          </div>
        </Cell>
        <Cell
          displayValue={14}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            14
          </div>
        </Cell>
        <Cell
          displayValue={15}
          handleCellclick={[Function]}
        >
          <div
            className="cell"
            onClick={[Function]}
          >
            15
          </div>
        </Cell>
        <Cell
          displayValue={0}
          handleCellclick={[Function]}
        >
          <div
            className="cell cell-empty"
            onClick={[Function]}
          />
        </Cell>
        <div />
        <Overlay
          displayText="WIN"
        >
          <div
            className="overlay"
          >
            WIN
          </div>
        </Overlay>
      </div>
    </Grid>
    <button
      className="button-reset"
      onClick={[Function]}
    >
      RESET
    </button>
  </main>
</App>
`;
